### ISSUE TITLE
json writer broken on some objects?

### ISSUE THREAD

**Issue by erg** [factor/factor#1189]
Should we allow ``>json`` to convert keys into strings if they are not already?

```factor
! Works
H{ { "1" 2 } { "3" 4 } } >json print
! {"1":2,"3":4}

! Fails in ``jsvar-encode`` because it treats keys as sequences
H{ { 1 2 } { 3 4 } } >json print
```

**Comment by jonenst** [comment]
Here's a patch for writing escaped unicode characters > 0x10000
```diff
commit 96004a9ff495dfad400f3c51707959c4a37c89b0
Author: Jon Harper <jon.harper87@gmail.com>
Date:   Fri Jan 2 19:30:26 2015 +0100

    JSON: writer, allow escaped unicode > 0x10000

diff --git a/basis/json/writer/writer-tests.factor b/basis/json/writer/writer-tests.factor
index 8f46af1..4c3d563 100644
--- a/basis/json/writer/writer-tests.factor
+++ b/basis/json/writer/writer-tests.factor
@@ -67,3 +67,5 @@ TUPLE: person first-name age ;
     "\0\x01\x02\x03\x04\x05\x06\a\b\t\n\v\f\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\e\x1c\x1d\x1e\x1f"
     >json
 ] unit-test
+
+{ "\"\\ud834\\udd1e\"" } [ t json-escape-unicode? [ "𝄞" >json ] with-variable ] unit-test
diff --git a/basis/json/writer/writer.factor b/basis/json/writer/writer.factor
index ec65429..55aca17 100644
--- a/basis/json/writer/writer.factor
+++ b/basis/json/writer/writer.factor
@@ -2,7 +2,8 @@
 ! See http://factorcode.org/license.txt for BSD license.
 USING: accessors ascii assocs combinators fry hashtables io
 io.streams.string json kernel locals math math.parser mirrors
-namespaces sequences strings tr words ;
+namespaces sequences strings tr words io.encodings.utf16.private
+formatting ;
 IN: json.writer
 
 SYMBOL: json-allow-fp-special?
@@ -39,6 +40,22 @@ M: t stream-json-print
 M: json-null stream-json-print
     [ drop "null" ] [ stream-write ] bi* ;
 
+: json-print-generic-escape-surrogate-pair ( stream char -- stream )
+    0x10000 - [ encode-first ] [ encode-second ] bi
+    "\\u%x%x\\u%x%x" sprintf over stream-write ;
+
+: json-print-generic-escape-bmp ( stream char -- stream )
+    "\\u" pick stream-write
+    >hex 4 CHAR: 0 pad-head
+    over stream-write ;
+
+: json-print-generic-escape ( stream char -- stream )
+    dup 0xffff > [
+        json-print-generic-escape-surrogate-pair
+    ] [
+        json-print-generic-escape-bmp
+    ] if ;
+
 M: string stream-json-print
     CHAR: " over stream-write1 swap [
         {
@@ -63,10 +80,7 @@ M: string stream-json-print
                     { [ dup control? ] [ t ] }
                     [ json-escape-unicode? get ]
                 } cond [
-                    dup 0xffff > [ json-error ] when
-                    "\\u" pick stream-write
-                    >hex 4 CHAR: 0 pad-head
-                    over stream-write
+                    json-print-generic-escape
                 ] [
                     over stream-write1
                 ] if
```

**Comment by jonenst** [comment]
Also, some things I plan to do, what do you think ?
   * better error reporting (we currently only have "json-error"), it could easily be more description (ie json-special-fp-error: value 1/0. ; other errors for other cases)
   * more output variables: don't always escape u2028 an u2029
   * fix escape-unicode? (I think it escapes everyting right now)
   * share same errors between reader and writer
   * share errors between reader and writer (for ex special-fp-error)

I have more stuff in mind, but do you think of anything else ?

**Comment by mrjbq7** [comment]
Patched in 1ec43d791347f16a817ef59fa506ab3a46f42ab2, thanks!

As for the other stuff, that all seems like good improvements!

**Comment by jonenst** [comment]
I manually parsed all the inputs from the jansson test suite "valid" and it looked good (except for "real-underflow" which hanged factor with "123e-10000000", see #1356 )
```factor
"/tmp/jansson/test/suites/valid/"
    [ directory-tree-files [ 5 short tail* "input" = ] filter ]
    [ [ prepend ] curry map ] bi
[ utf8 file-contents json> . ] each
```
I'll look into importing those tests as unit-tests in factor.

**Comment by erg** [comment]
Can we write some code to git clone/download and somehow test the jansson test suite? Or just port some of the interesting tests?

**Comment by bpollack** [comment]
All right, I've gone through the entire Jasson suite by hand and added a couple minor cases we were missing. I think this can be closed in good faith once the PR lands.

**Comment by mrjbq7** [comment]
Merged #1543.
